python/semantic_kernel/connectors/memory/postgres/postgres_collection.py

```diff
@@ -0,0 +1,412 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+import logging
+import sys
+from collections.abc import Sequence
+from typing import Any, ClassVar, TypeVar
+
+from pydantic import PrivateAttr
+
+from semantic_kernel.connectors.memory.postgres.postgres_settings import PostgresSettings
+from semantic_kernel.connectors.memory.postgres.utils import (
+    convert_dict_to_row,
+    convert_row_to_dict,
+    get_vector_index_ops_str,
+    python_type_to_postgres,
+)
+from semantic_kernel.data.const import IndexKind
+from semantic_kernel.data.vector_store_model_definition import VectorStoreRecordDefinition
+
+if sys.version_info >= (3, 12):
+    from typing import override  # pragma: no cover
+else:
+    from typing_extensions import override  # pragma: no cover
+
+from psycopg import sql
+from psycopg.errors import DatabaseError
+from psycopg_pool import AsyncConnectionPool
+
+from semantic_kernel.connectors.memory.postgres.constants import DEFAULT_SCHEMA, MAX_DIMENSIONALITY
+from semantic_kernel.data.vector_store_record_collection import VectorStoreRecordCollection
+from semantic_kernel.data.vector_store_record_fields import VectorStoreRecordKeyField, VectorStoreRecordVectorField
+from semantic_kernel.exceptions.memory_connector_exceptions import (
+    MemoryConnectorException,
+    VectorStoreModelValidationError,
+)
+from semantic_kernel.kernel_types import OneOrMany
+from semantic_kernel.utils.experimental_decorator import experimental_class
+
+TKey = TypeVar("TKey", str, int)
+TModel = TypeVar("TModel")
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+@experimental_class
+class PostgresCollection(VectorStoreRecordCollection[TKey, TModel]):
+    """PostgreSQL collection implementation."""
+
+    connection_pool: AsyncConnectionPool | None = None
+    db_schema: str = DEFAULT_SCHEMA
+    supported_key_types: ClassVar[list[str] | None] = ["str", "int"]
+    supported_vector_types: ClassVar[list[str] | None] = ["float"]
+
+    _handle_pool_close: bool = PrivateAttr(False)
+    """Whether the collection should handle closing the pool. True if the pool was created by the collection."""
+
+    _settings: PostgresSettings = PrivateAttr()
+    """Postgres settings"""
+
+    def __init__(
+        self,
+        collection_name: str,
+        data_model_type: type[TModel],
+        data_model_definition: VectorStoreRecordDefinition | None = None,
+        connection_pool: AsyncConnectionPool | None = None,
+        db_schema: str = DEFAULT_SCHEMA,
+        env_file_path: str | None = None,
+        env_file_encoding: str | None = None,
+        settings: PostgresSettings | None = None,
```

Comment:
@moonbox3  This was anticipation for another connector I'm creating for AzureDBForPostgres. By passing the settings to the constructor, I'm able to subclass both this class and the PostgresSettings class to allow for a token generation function using Entra in lieu of a password. Without enabling settings to be passed up to the super class, there's no way to enable subtypes to be used as setting objects rather than the exact PostgresSetting type. The subtype was a way to have an alternate implementation of the `get_connection_args` method in the settings class, which pulls connection parameters from either the connection string or the settings; this allowed setting the Entra token generation callables into the password.

An alternate approach that I can use that better aligns with the way you outlined is to modify the current PostgresSettings to accept a callable for the password, and pass the parameters directly. That doesn't allow subclasses of the settings to be used, and will require duplication of parameters between settings and constructors, but will work in this particular case. I'll make that change.

Reply:
Ah, I was trying to work through this and caught another reason why holding onto the settings is important.

The way Redis is implemented, the client is created as part of the __init__. It is wrapped so that the connection can be closed during the `__del__` method via the RedisClientWrapper here: https://github.com/microsoft/semantic-kernel/blob/3e529d053b13511830500dfe1b212ab1d5df8353/python/semantic_kernel/connectors/memory/redis/utils.py#L129.

This doesn't allow the context manager methods of `__aenter__` and `__aexit__` to do connection management, and instead relies on object deletion to close connections. In the case of the psycopg client, this does not work (see https://github.com/microsoft/semantic-kernel/pull/8951#discussion_r1783445068 for more discussion).

In order for the connection to be managed via the context manager functions, the connection pool should be created and then closed and deleted in those async functions. To do this, I need to hold onto the settings that provide the connection parameters.

Note that the VectorStore itself is not an async context manager, so I force the user to pass in their own connection pool if using a VectorStore and not a VectorStoreRecordCollection directly, due to the inability to manage async connections correctly.

Thoughts?